### PR TITLE
build: add isolated J2CL sidecar scaffold

### LINKED ISSUE #900: Stand up the isolated J2CL sidecar build and SBT entrypoints
## Goal
Create the first isolated J2CL sidecar build for SupaWave and wire it into the existing SBT-first repo without cutting over any runtime traffic yet.

## Why now
The repo still has no `j2cl/` sidecar, no Maven wrapper, and no `j2cl-maven-plugin` configuration. The full migration plan already specifies the build shape, but none of that scaffold exists in the live tree yet.

## Source docs
- `docs/superpowers/plans/j2cl-full-migration-plan.md` sections 6.3 through 6.6
- `docs/j2cl-gwt3-decision-memo.md`
- `docs/j2cl-preparatory-work.md`

## Scope
- add `j2cl/pom.xml` with `com.vertispan.j2cl:j2cl-maven-plugin`
- add Maven wrapper files under `j2cl/`
- add a minimal `j2cl/src/main/webapp/**` host page and assets for a sandbox bundle
- configure explicit plugin profiles/executions for:
  - sandbox or search-sidecar (`BUNDLE_JAR`, sourcemaps on)
  - debug-single-project (`BUNDLE`, sourcemaps on)
  - production (`ADVANCED_OPTIMIZATIONS`, sourcemaps retained as artifacts)
- add SBT tasks that invoke the sidecar build in a reproducible way
- keep the output isolated under a dedicated `war/j2cl*/**` path

## Non-goals
- no app-shell cutover
- no transport rewrite yet
- no search/UI migration yet

## Acceptance Criteria
- the repo contains the new `j2cl/` sidecar scaffold and wrapper files
- an SBT task such as `j2clSandboxBuild` produces a valid J2CL JavaScript bundle
- a corresponding test task such as `j2clSandboxTest` is wired and documented, even if the initial test set is minimal
- the sidecar build does not replace the existing GWT bundle or change default runtime behavior
- the output layout is explicit and compatible with later dual-bundle work

### PR DESCRIPTION
**Pull request by vega113**
Closes #900

## Summary
- add an isolated `j2cl/` Maven sidecar project with wrapper-backed sandbox, search-sidecar, debug, and production build profiles
- add opt-in SBT entrypoints for the sidecar build/test flow without changing the legacy GWT runtime bootstrap
- serve isolated sidecar assets under `/j2cl-search/*`, `/j2cl-debug/*`, and `/j2cl/*` while keeping `/` and `/webclient/*` on the current GWT path
- add the issue-specific implementation plan and changelog fragment for the new sidecar scaffold

## Verification
- `./j2cl/mvnw -f j2cl/pom.xml -Psearch-sidecar -q test`
- `./j2cl/mvnw -f j2cl/pom.xml -Pdebug-single-project -q package`
- `./j2cl/mvnw -f j2cl/pom.xml -Pproduction -q package`
- `sbt "j2clSandboxBuild" "j2clSandboxTest" "j2clSearchBuild" "j2clSearchTest"`
- `sbt compile test`
  - passed: `Passed: Total 2289, Failed 0, Errors 0, Passed 2287, Skipped 2`
- `python3 scripts/assemble-changelog.py`
- `python3 scripts/validate-changelog.py --changelog wave/config/changelog.json`
- `sbt Universal/stage`
- `bash scripts/worktree-boot.sh --port 9900`
- `PORT=9900 JAVA_OPTS='-Djava.util.logging.config.file=/Users/vega/devroot/worktrees/issue-900-j2cl-sidecar-build/wave/config/wiab-logging.conf -Djava.security.auth.login.config=/Users/vega/devroot/worktrees/issue-900-j2cl-sidecar-build/wave/config/jaas.config' bash scripts/wave-smoke.sh start`
- `PORT=9900 bash scripts/wave-smoke.sh check`
  - `ROOT_STATUS=200`
  - `HEALTH_STATUS=200`
  - `WEBCLIENT_STATUS=200`
- `curl -sS -I http://127.0.0.1:9900/webclient/webclient.nocache.js`
  - `200 OK`
- `curl -sS -I http://127.0.0.1:9900/j2cl-search/index.html`
  - `200 OK`
- `curl -sS -I http://127.0.0.1:9900/j2cl/index.html`
  - `200 OK`
- browser verification on `http://127.0.0.1:9900/`, `http://127.0.0.1:9900/j2cl-search/index.html`, and `http://127.0.0.1:9900/j2cl/index.html`
  - root `/` still showed the legacy SupaWave landing/sign-in flow
  - `/j2cl-search/index.html` loaded the isolated sandbox page with the `search-sidecar` summary
  - `/j2cl/index.html` loaded the isolated sandbox page with the `production` summary
- `PORT=9900 bash scripts/wave-smoke.sh stop`

## Traceability
- Worktree: `/Users/vega/devroot/worktrees/issue-900-j2cl-sidecar-build`
- Plan: `docs/superpowers/plans/2026-04-18-issue-900-j2cl-sidecar-build.md`
- Local verification record: `journal/local-verification/2026-04-19-issue-900-j2cl-sidecar-build.md`

## Notes
The legacy GWT runtime remains the active root/bootstrap path in this PR. The new J2CL outputs are sidecar-only scaffolding for later issues.

## Summary by CodeRabbit

* **New Features**
  * Opt-in J2CL Maven sidecar that builds isolated browser artifacts and a lightweight sidecar UI served at /j2cl-search/, /j2cl-debug/, and /j2cl/.
  * New build and test tasks to produce and verify sidecar and search artifacts.
  * Lightweight sandbox entrypoint, host page and stylesheet for sidecar preview.

* **Chores**
  * Added implementation plan, changelog entry, Maven wrapper, .gitignore entries, cleanup of generated sidecar outputs, and a smoke test.

### CODE REVIEW

File: j2cl/src/main/webapp/index.html [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        if (pathname.indexOf("/j2cl-search/") !== -1) {
+          return "search-sidecar";
+        }
+        if (pathname.indexOf("/j2cl-debug/") !== -1) {
+          return "debug-single-project";
+        }
+        if (pathname.indexOf("/j2cl/") !== -1) {
+          return "production";
+        }
+        return "sidecar";
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/webapp/index.html:37]: Fixed. Added exact-path equality checks alongside the indexOf variants for all three roots.

File: j2cl/src/test/java/org/waveprotocol/box/j2cl/sandbox/SandboxBuildSmokeTest.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+
+@J2clTestInput(SandboxBuildSmokeTest.class)
+public class SandboxBuildSmokeTest {
+  @Test
+  public void renderSummaryIncludesRequestedMode() {
+    Assert.assertEquals(
+        "Profile search-sidecar writes isolated assets without changing the root runtime bootstrap.",
+        SandboxEntryPoint.renderSummary("search-sidecar"));
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/test/java/org/waveprotocol/box/j2cl/sandbox/SandboxBuildSmokeTest.java]: Fixed. Test now uses `assertTrue(summary.contains(mode))` — profile-agnostic and easier to maintain as the summary text evolves.

File: j2cl/.mvn/wrapper/maven-wrapper.properties [hunk 3]
```diff
@@ -0,0 +1,2 @@
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.9.5/apache-maven-3.9.5-bin.zip
+wrapperUrl=https://repo.maven.apache.org/maven2/org/apache/maven/wrapper/maven-wrapper/3.3.4/maven-wrapper-3.3.4.jar
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/.mvn/wrapper/maven-wrapper.properties:2]: Fixed. Fetched the actual SHA-256 from Maven Central (downloaded the jar and computed locally) and added `wrapperSha256Sum=4e2fbf6554bc8a4702cdfdd3bef464f423393d784ddbb037216320ce55d5e4e1`.

File: j2cl/mvnw.cmd [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+if exist "%USERPROFILE%\mavenrc_post.bat" call "%USERPROFILE%\mavenrc_post.bat"
+if exist "%USERPROFILE%\mavenrc_post.cmd" call "%USERPROFILE%\mavenrc_post.cmd"
+:skipRcPost
+
+@REM pause the script if MAVEN_BATCH_PAUSE is set to 'on'
+if "%MAVEN_BATCH_PAUSE%"=="on" pause
+
+if "%MAVEN_TERMINATE_CMD%"=="on" exit %ERROR_CODE%
+
+cmd /C exit /B %ERROR_CODE%
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/mvnw.cmd:206]: Fixed. Added `.gitattributes` enforcing CRLF on `j2cl/mvnw.cmd` and re-normalized the file with `git add --renormalize`.

File: wave/src/jakarta-overrides/java/org/waveprotocol/box/server/rpc/ServerRpcProvider.java [hunk 5]
```diff
@@ -537,6 +543,15 @@ public void startWebSocketServer(final Injector injector) {
...
                 webclientResource = resolveResource(ResourceFactory.of(context), "webclient");
             }
+            if ((j2clSearchResource == null || !j2clSearchResource.exists()) && resourceBases != null) {
+                j2clSearchResource = resolveResource(ResourceFactory.of(context), "j2cl-search");
+            }
+            if ((j2clDebugResource == null || !j2clDebugResource.exists()) && resourceBases != null) {
+                j2clDebugResource = resolveResource(ResourceFactory.of(context), "j2cl-debug");
+            }
+            if ((j2clResource == null || !j2clResource.exists()) && resourceBases != null) {
+                j2clResource = resolveResource(ResourceFactory.of(context), "j2cl");
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/jakarta-overrides/java/org/waveprotocol/box/server/rpc/ServerRpcProvider.java]: Fixed. Removed the `resolveResource` fallback for all three j2cl routes. The fallback's relative-path resolution via `Paths.get("j2cl")` was finding the repo-root Maven source directory when `war/j2cl/` hadn't been built, so the servlet would mount against `pom.xml`/`src/` instead of generated assets. J2CL servlets are now only registered when the artifact directory explicitly exists under `baseResource`.

File: build.sbt [hunk 6]
```diff
@@ -1309,3 +1357,6 @@ cleanFiles += baseDirectory.value / "war" / "webclient"
 cleanFiles += baseDirectory.value / "war" / "org"
 cleanFiles += baseDirectory.value / "war" / "WEB-INF"
 cleanFiles += baseDirectory.value / "war-dev"
+cleanFiles += baseDirectory.value / "war" / "j2cl-search"
+cleanFiles += baseDirectory.value / "war" / "j2cl-debug"
+cleanFiles += baseDirectory.value / "war" / "j2cl"
```
[the remark this answers is not in the document]
Comment (vega113) [reply build.sbt:1362]: Addressed in commit `121f1aa78` on the PR branch. The branch now ignores `war/j2cl/`, `war/j2cl-search/`, and `war/j2cl-debug/` so running the sidecar build does not leave generated bundles untracked in `git status`.

File: j2cl/src/main/webapp/index.html [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      }
+
+      function renderFallback(mode) {
+        var summary = document.querySelector("[data-sidecar-summary]");
+        if (summary) {
+          summary.textContent = "Profile " + mode + " writes isolated assets without changing the root runtime bootstrap.";
+        }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/webapp/index.html:48]: Fixed. Added `renderLoadError()` — when `attemptsRemaining` hits 0 without finding `WaveSandboxEntryPoint`, the summary now shows a red error message ("ERROR: sidecar bundle failed to load – WaveSandboxEntryPoint not defined.") so a missing/broken bundle is clearly distinguishable from a successful mount.

[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/webapp/index.html:48]: Addressed in `99f6edcd0`. The fallback copy no longer matches the successful mount summary: it now stays in an explicit placeholder/error state until `WaveSandboxEntryPoint.mount(...)` replaces the card, so a missing sidecar bundle is visibly distinguishable during verification.

**Comment by vega113** [comment]
## Merge Conflict Resolution

This PR had conflicts with main that have been resolved.

### Conflicting Files
- `j2cl/.mvn/wrapper/maven-wrapper.properties`
- `j2cl/src/main/webapp/index.html`
- `j2cl/src/test/java/org/waveprotocol/box/j2cl/sandbox/SandboxBuildSmokeTest.java`

### Resolution Summary

All three were add/add conflicts caused by main absorbing some of the sidecar scaffold via `6a41a140` (_fix(j2cl): show explicit error when sidecar bundle fails to mount_) before this PR was merged.

**maven-wrapper.properties:** Kept main's version with the `wrapperSha256Sum` security checksum (the PR's original version omitted it).

**index.html:** Kept main's version with exact-path matching (`pathname === "/j2cl-search"`) and the explicit `renderLoadError()` function — both improvements added in the fix commit.

**SandboxBuildSmokeTest.java:** Kept main's version with both assertions (mode inclusion test + whitespace-trimming edge case), which provides fuller coverage.

### Verification
- [x] No conflict markers remain
- [x] Rebase completed cleanly (6/6 commits applied)
- [x] Force-pushed to `issue-900-j2cl-sidecar-build`

**Comment by vega113** [comment]
Fixed. When `addJ2clServlet` is called and the artifact directory is absent, it now registers an explicit servlet that returns `404 Not Found` (with a descriptive message) instead of returning early and letting the request fall through to the root app. This makes route-level health checks reliable — `curl -I /j2cl/index.html` will correctly fail when the sidecar hasn't been built.

**Comment by vega113** [comment]
## Merge Conflict Resolution

Conflicts with main have been resolved via rebase.

### Conflicting File
- `wave/src/jakarta-overrides/java/org/waveprotocol/box/server/rpc/ServerRpcProvider.java`

### Resolution Summary

**Conflict 1 — J2CL resource fallback block:**
Kept main's version: removed the `resolveResource` fallback for j2cl servlet roots (fix from `ba0bf9146`). The fallback was incorrect because it could match the `j2cl/` Maven source directory when sidecar artifacts haven't been built.

**Conflict 2 — `addJ2clServlet` absent-resource handling:**
Kept main's version: registers an explicit 404 servlet when sidecar artifacts are absent (fix from `ba0bf9146`), instead of silently returning and letting requests fall through to the root app.

One commit (`b0a16a89`) was dropped during rebase — it was already upstream on main.

File: wave/src/jakarta-overrides/java/org/waveprotocol/box/server/rpc/ServerRpcProvider.java [hunk 8]
```diff
@@ -829,6 +846,24 @@ public ServletHolder addServlet(String urlPattern, Class<?> servlet) {
...
 
+    private static void addJ2clServlet(
+            ServletContextHandler context,
+            String holderName,
+            @Nullable Resource resource,
+            String displayPath,
+            String mapping) {
+        if (resource == null || !resource.exists()) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/jakarta-overrides/java/org/waveprotocol/box/server/rpc/ServerRpcProvider.java:871]: Fixed in `ba0bf9146`. `addJ2clServlet` now registers an inline 404 servlet for each j2cl path when the artifact directory is absent, so requests return SC_NOT_FOUND with a clear message rather than falling through to the root app.

[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/jakarta-overrides/java/org/waveprotocol/box/server/rpc/ServerRpcProvider.java:871]: Fixed in the rebased branch. `addJ2clServlet(...)` now keeps the explicit 404 handler when `war/j2cl*` artifacts are absent, so `/j2cl*` no longer falls through to the root app. The branch was rebased onto `ba0bf9146` and force-pushed at `d35d4c62d`.
